Modules/_sqlite/connection.c

```diff
@@ -432,48 +436,75 @@ free_callback_contexts(pysqlite_Connection *self)
 static void
 remove_callbacks(sqlite3 *db)
 {
+    if (db == NULL) {
+        return;
+    }
     sqlite3_trace_v2(db, SQLITE_TRACE_STMT, 0, 0);
     sqlite3_progress_handler(db, 0, 0, (void *)0);
     (void)sqlite3_set_authorizer(db, NULL, NULL);
 }
 
-static void
+static int
 connection_close(pysqlite_Connection *self)
 {
     if (self->db) {
         if (self->autocommit == AUTOCOMMIT_DISABLED &&
             !sqlite3_get_autocommit(self->db))
         {
-            /* If close is implicitly called as a result of interpreter
-             * tear-down, we must not call back into Python. */
-            if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
-                remove_callbacks(self->db);
+            if (connection_exec_stmt(self, "ROLLBACK") < 0) {
+                return -1;
```

Comment:
I would prefer to execute the following code anyway, and return -1 on error.

Reply:
Hm, yes that is probably a safer option. We really want the database to be closed.